File: leetcode/1381 Design a Stack With Increment Operation/solution.java
class CustomStack { // 将Increment操作复杂度由O(k)降到O(1)！！！
    int[] stack;
    int top; // 记录栈顶位置！！！
    int[] add; // 记录增量！！！

    public CustomStack(int maxSize) {
        this.stack = new int[maxSize];
        this.add = new int[maxSize];
        this.top = -1;
    }
    
    public void push(int x) {
        if (top != stack.length-1) {
            top++;
            stack[top] = x;
        }
    }
    
    public int pop() {
        if (top == -1) return -1;
        int result = stack[top] + add[top];
        if (top != 0) {
			add[top-1] += add[top]; // 栈顶元素被弹出，增量需要往下传递！！！
        }
        add[top] = 0;
        top--;
        return result;
    }
    
    public void increment(int k, int val) { // 调用时只改变栈顶元素的增量！！！
        int limit = Math.min(k-1, top);
        if (limit >= 0) {
			add[limit] += val;
        }
    }
}

/**
 * Your CustomStack object will be instantiated and called as such:
 * CustomStack obj = new CustomStack(maxSize);
 * obj.push(x);
 * int param_2 = obj.pop();
 * obj.increment(k,val);
 */

File: leetcode/1381 Design a Stack With Increment Operation/solution.go
type CustomStack struct { // 定义结构体！！！
	stack []int
	top int
}

func Constructor(maxSize int) CustomStack { // 构造器返回实例！！！
	return CustomStack{
		stack : make([]int, maxSize),
		top : -1,
	}
}

func (this *CustomStack) Push(x int) {
	if this.top != len(this.stack)-1 {
		this.top++
        this.stack[this.top] = x
	}
}

func (this *CustomStack) Pop() int {
    if this.top == -1 {
        return -1
    }
    this.top--
    return this.stack[this.top+1]
}

func (this *CustomStack) Increment(k int, val int) { // Increment操作复杂度由O(k)！！！
    limit := k
	if k > this.top+1 {
        limit = this.top+1
	}
	for i:=0; i<limit; i++ {
		this.stack[i] += val
	}
}

/**
 * Your CustomStack object will be instantiated and called as such:
 * obj := Constructor(maxSize);
 * obj.Push(x);
 * param_2 := obj.Pop();
 * obj.Increment(k,val);
 */